### PR TITLE
readpass_file

### PR DESCRIPTION
**Pull request by gperciva**

### CODE REVIEW

**Comment by gperciva** [comment]
Draft PR to discuss exactly what this function should do.  Things like:
- is the maximum length of 256 chars ok?
- should we strip a final `\n` and `\r\n`?
- is it ok to bail if there's any other `\n` or `\r` characters in the line?
- should we support an empty file / file only containing a newline, or not?

I found it useful to construct a whole bunch of `good-*.pass` and `bad-*.pass` files to clarify a bunch of edge cases.

**Comment by gperciva** [comment]
TODOs:
- add a test containing exactly the maximum number of chars.
- add a test containing exactly 1 more than the maximum number of chars.
- check on more OSes to see if the `fopen: No such file or directory" message is consistent, or whether I should change that test to a `grep` instead.

File: util/readpass_file.c [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+readpass_file(char ** passwd, const char * filename)
+{
+	struct stat sb;
+	FILE * fp;
+	char passbuf[MAXPASSLEN];
+	size_t len;
+	size_t eol_pos;
+
+	/* Stat the file. */
+	if (stat(filename, &sb)) {
```
Comment (cperciva) [inline util/readpass_file.c:33]: There's a race condition here if the file is replaced between the stat call and the fopen.  I would fopen first and then use fstat(fileno(f)).

Comment (cperciva) [reply util/readpass_file.c:33]: ... or for that matter just fopen and fgets, without needing to stat the file at all.

**Comment by gperciva** [comment]
Switched to fgets, rebased the whole thing.

File: util/readpass_file.c [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+ * Read a passphrase from ${filename} and return it as a malloced
+ * NUL-terminated string via ${passwd}.  Produce an error if the file is 2048
+ * characters or more, or if it contains any newline \n or \r\n characters
+ * other than at the end of the file.  Do not include the \n or \r\n
+ * characters in the passphrase.
+ */
+int
+readpass_file(char ** passwd, const char * filename)
+{
+	FILE * fp;
```
Comment (cperciva) [inline util/readpass_file.c:23]: Why `fp`?  Normally we just call a file `f`...

File: util/readpass_file.c [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+			warnp("fread(\"%s\")", filename);
+			goto err2;
+		} else {
+			/* We have a 0-byte password. */
+			passbuf[0] = '\0';
+		}
+	}
+
+	/* Bail if there's a second line. */
+	if ((fgets(secondline, 2, fp)) != NULL) {
```
Comment (cperciva) [inline util/readpass_file.c:45]: Why not `fgetc`?  That way you don't need a second buffer.

File: util/readpass_file.c (hunk 2 above)
Comment (gperciva) [reply util/readpass_file.c:23]: file pointer?  I hate one-character variables other than `i` and `j`.

... that said, I admit that you like `f` so I should follow that.

Comment (gperciva) [reply util/readpass_file.c:23]: Also, easier for searching -- not many words contain `fp`, but tons contain `f`.

File: util/readpass_file.c (hunk 3 above)
Comment (gperciva) [reply util/readpass_file.c]: done

File: util/readpass.h [hunk 4]
```diff
@@ -20,4 +20,14 @@
  */
 int readpass(char **, const char *, const char *, int);
 
+/**
+ * readpass_file(passwd, filename):
+ * Read a passphrase from ${filename} and return it as a malloced
+ * NUL-terminated string via ${passwd}.  Produce an error if the file is 2048
```
Comment (cperciva) [inline util/readpass.h:26]: "Produce an error" seems like it should be "Print an error and fail".

File: util/readpass_file.c [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+			goto err2;
+		} else {
+			/* We have a 0-byte password. */
+			passbuf[0] = '\0';
+		}
+	}
+
+	/* Bail if there's a second line. */
+	if (fgetc(f) != EOF) {
+		warn0("more than 1 line in %s", filename);
```
Comment (cperciva) [inline util/readpass_file.c:45]: We could get here if there's a single overly long line, I think?

File: util/readpass_file.c [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+ */
+int
+readpass_file(char ** passwd, const char * filename)
+{
+	FILE * f;
+	char passbuf[MAXPASSLEN];
+
+	/* Open the file. */
+	if ((f = fopen(filename, "r")) == NULL) {
+		warnp("fopen(\"%s\")", filename);
```
Comment (cperciva) [inline util/readpass_file.c:28]: I think I usually omit the `\"` in warning strings like this.

File: util/readpass_file.c [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+/**
+ * readpass_file(passwd, filename):
+ * Read a passphrase from ${filename} and return it as a malloced
+ * NUL-terminated string via ${passwd}.  Produce an error if the file is 2048
+ * characters or more, or if it contains any newline \n or \r\n characters
+ * other than at the end of the file.  Do not include the \n or \r\n
+ * characters in the passphrase.
+ */
+int
```
Comment (cperciva) [inline util/readpass_file.c:20]: Why not return a `char *`?

File: util/readpass_file.c (hunk 5 above)
Comment (gperciva) [reply util/readpass_file.c:45]: yuck, yes.  That makes for a more confusing warning, but I suppose that the code clarity from switching to `fgetc` is still worth it.

File: util/readpass_file.c (hunk 7 above)
Comment (gperciva) [reply util/readpass_file.c:20]: to match:
```
int readpass(char **, const char *, const char *, int);
```

Comment (cperciva) [reply util/readpass_file.c:20]: Hmm, ok.

**Comment by cperciva** [comment]
Looks ok, please clean up the history.

**Comment by gperciva** [comment]
Rebased, ready for merge.
